### PR TITLE
Verification and Gas control

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/789

### CODE REVIEW

File: neo/SmartContract/ApplicationEngine.cs [hunk 1]
```diff
@@ -12,26 +12,22 @@ public partial class ApplicationEngine : ExecutionEngine
...
-        public const long GasFree = 0;
-        private readonly long gas_amount;
-        private readonly bool testMode;
         private readonly RandomAccessStack<UInt160> hashes = new RandomAccessStack<UInt160>();
         private readonly List<NotifyEventArgs> notifications = new List<NotifyEventArgs>();
         private readonly List<IDisposable> disposables = new List<IDisposable>();
 
         public TriggerType Trigger { get; }
         public IVerifiable ScriptContainer { get; }
+        public IExecutionControl Control { get; }
```
Comment (shargon) [inline neo/SmartContract/ApplicationEngine.cs]: This could be in `neo-vm` by default

**Comment by erikzhang** [comment]
What problem does this refactoring solve?

**Comment by shargon** [comment]
We are able to ban opcodes in verification, and max amount of instructions to execute

https://github.com/neo-project/neo/blob/ca8de05b4971d59df63a7f51c56b70555063ab9d/neo/SmartContract/VerificationControl.cs#L6-L23

**Comment by erikzhang** [comment]
I think limit the max steps is useless. Limiting the max gas is enough.

**Comment by shargon** [comment]
And the opcodes, like syscalls?

**Comment by erikzhang** [comment]
Without SYSCALL, the verification script can't verify the signatures. There is no reason to limit it.

File: neo/SmartContract/VerificationControl.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+﻿using Neo.VM;
+using System.Collections.Generic;
+
+namespace Neo.SmartContract
+{
+    public class VerificationControl : GasControl
+    {
+        private readonly static List<OpCode> _bannedOpcodes = new List<OpCode>(new OpCode[] { /*OpCode.NOP*/ });
```
Comment (igormcoelho) [inline neo/SmartContract/VerificationControl.cs]: Its in a good direction shargon, however needs a little more complication here. To ban backwards jumps we need to analyse the code and its first parameter... so, a callback returning true/false is better than a static list.
I've been thinking and another option would be to have different JMPF that can only jump forward. If we do this, its easier to ban. 3 more: JMPF, JMPFIF and JMPFIFNOT . A callback can avoid this and give more control on used opcodes.

**Comment by vncoelho** [comment]
I really do not have the perfect understanding that you @igormcoelho, @lightszero, @erikzhang and @shargon have been sharing over these topics.

However, biased or not, I am in favor of things that my brother @igormcoelho  proposes, related to not being turing complete and remove the jumps in some special cases.
We could have both system working together. Those that want limited execution and those that do not want it.

After that, we should also please come back to that discussion about `op_eval`. Let's turn it to a `SYSCALL` and well and allow us to play with this! :dancer:

File: neo/SmartContract/GasControl.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    public partial class GasControl : IExecutionControl
+    {
+        public const long GasFree = 0;
+
+        private readonly long gas_amount;
+        private readonly bool testMode;
+
+        public long GasConsumed { get; private set; } = 0;
+
+        public GasControl(long gas = 0, bool testMode = false)
```
Comment (vncoelho) [inline neo/SmartContract/GasControl.cs]: Perfect, perfect, @shargon.

The refactor is precise!
It is much more readable.

I agree with it.

The other discussions out of the refactor and the class could be moved to other PR.

Maybe we can rename `GasControl` -> `GasManager` or `GasTuringCompleteManager` or `GasTuringIncompleteManager`...aehuaheuaheua
Turing is just for fun. Maybe `GasManager` or `GasLimitsManager`

File: neo/SmartContract/VerificationControl.cs (hunk 2 above)
Comment (vncoelho) [reply neo/SmartContract/VerificationControl.cs]: This gonna be an interesting thing!

Comment (vncoelho) [reply neo/SmartContract/VerificationControl.cs]: Do you want to handle these bannedOpcodes in this PR or split in two?

Comment (vncoelho) [reply neo/SmartContract/VerificationControl.cs]: @igormcoelho

**Comment by vncoelho** [comment]
Can we move with this PR? It has a nice refactor, maybe split this or leave this template and later adjust.

**Comment by vncoelho** [comment]
@shargon, maybe split this PR in Refactoring and the feature. I like the refactor.

**Comment by vncoelho** [comment]
ping @shargon...ahueahuea

**Comment by shargon** [comment]
If all are agree i can fix the conflicts otherwise i will close it. Now i think that is not needed because we only check gas

**Comment by erikzhang** [comment]
I think it's not needed anymore.

**Pull request closed without merging**